### PR TITLE
Use k8s utils/pointer instead of Azure autorest/to

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Replaces usage of the `github.com/Azure/go-autorest/autorest/to` package with `k8s.io/utils/pointer` and a couple small helper funcs. This makes forward progress toward removing usage of Azure/go-autorest, which will soon be unsupported.

In general, the changes here consist of:

| autorest/to      | utils/pointer |
| ----------- | ----------- |
| to.StringPtr      | pointer.String       |
| to.Float64Ptr   | pointer.Float64        
| ...etc   |         |
| to.String  | pointer.StringDeref        |
| to.Int32   | pointer.Int32Deref       |
| ...etc  |         |
| to.StringSlicePtr   | &        |
| to.StringSlice   | (local) azure.StringSlice        |
| to.StringMapPtr   | (local) azure.StringMapPtr      |

See kubernetes-sigs/cloud-provider-azure#2983 for comparison.

**Which issue(s) this PR fixes**:

Refs #2974

**Special notes for your reviewer**:

This changeset is huge; I'd be glad to break it up into multiple PRs if that helps with reviewing.

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by mboersma** [comment]
/test pull-cluster-api-provider-azure-e2e-exp

AKS provisioning failure, unrelated to this code change.

**Comment by mboersma** [comment]
/hold

Let's wait until #2959 merges first.

**Review by CecileRobertMichon — COMMENTED** [review]
/assign @willie-yao @Jont828

File: api/v1beta1/azurecluster_default_test.go
Comment (willie-yao) [inline api/v1beta1/azurecluster_default_test.go:142]: I see that there are some uses of `pointer.Int32Ptr` and `pointer.Int32Deref`. The PR description mentions that `pointer.Int32Deref` is being used to replace `to.Int32Ptr`. Will this be a problem? I noticed that `pointer.Int32Deref` is used for cases where the int might be nil, so I think it is the correct usage.

Comment (mboersma) [reply api/v1beta1/azurecluster_default_test.go:142]: > `pointer.Int32Deref` is being used to replace `to.Int32Ptr`

Actually it's like this:

`to.Int32Ptr` -> `pointer.Int32` (return a pointer / reference)
`to.Int32` -> `pointer.Int32Deref` (return an int32 or 0)

Those changes *should* preserve identical behavior.

**Comment by mboersma** [comment]
Rebased after #3093 changed go.mod.

**Review by willie-yao — APPROVED** [review]
/lgtm

File: api/v1beta1/azurecluster_default_test.go
Comment (CecileRobertMichon) [reply api/v1beta1/azurecluster_default_test.go:142]:
```
// Int32Ptr is a function variable referring to Int32.
// Deprecated: Use Int32 instead.
var Int32Ptr = Int32 // for back-compat
```

seems we should be using `Int32` though? `Int32Ptr` is deprecated

Comment (mboersma) [reply api/v1beta1/azurecluster_default_test.go:142]: Oops, yes, that's a mistake that crept in through a rebase: `pointer.Int32` is preferred now. I'll update.

**Comment by mboersma** [comment]
I had to rebase again for changes in `main`, and I replaced some pre-existing usages of `pointer.Int32Ptr` with `pointer.Int32` and the like.

**Comment by mboersma** [comment]
This PR is old enough now that `prow` is holding it hostage because the `pull-cluster-api-provider-azure-e2e-exp` job didn't pass (and it never will because it's the `-aks` job now). So it may have to be admin-merged.

**Pull request merged**
